### PR TITLE
[Bugfix] Fix Dense module loading for sentence-transformers embedding models (simplified V2)

### PR DESCRIPTION
**Pull request by FFFfff1FFFfff**
**Purpose:**
This PR adds automatic support for Sentence-Transformers Dense projection layers in vLLM, enabling proper handling of models that require dimension transformation (e.g., 1024→1792) during embedding generation.

**Resolves the following issues:**

- Missing Dense projection functionality for ST models in vLLM
- Incorrect output dimensions (1024 instead of expected 1792 for models like TencentBAC/Conan-embedding-v1)
- Ensures numerical consistency with HuggingFace Sentence-Transformers implementation

**Key Modifications**

- vllm/model_executor/models/adapters.py: Added _load_st_projector() and weight loading logic
- vllm/model_executor/layers/pooler.py: Enhanced EmbeddingPoolerHead with projector support
- vllm/model_executor/models/bert.py: Integrated projector into BERT embedding models
- vllm/model_executor/layers/activation.py: Added tanh and sigmoid activation functions
- vllm/transformers_utils/config.py: Added get_hf_file_bytes() utility
- tests/models/language/pooling/test_st_projector.py: Added comprehensive test

**New Version Improvements**

- Simplified testing
- Reused vLLM methods
- Minimal code changes

**Test**

`python -m pytest tests/models/language/pooling/test_st_projector.py -v`

**Test Result**

`tests/models/language/pooling/test_st_projector.py::test_embed_models_mteb[model_info0] PASSED`

### CODE REVIEW

File: tests/models/language/pooling/mteb_utils.py [hunk 1]
```diff
@@ -182,8 +182,13 @@ def mteb_test_embed_models(hf_runner,
...
             assert model_info.architecture in model_config.architectures
-        assert (model_config._model_info.default_pooling_type ==
-                model_info.default_pooling_type)
+        if hasattr(model_config,
+                   'pooler_config') and model_config.pooler_config:
+            actual_pooling = model_config.pooler_config.pooling_type
+            assert actual_pooling == model_info.default_pooling_type
+        else:
+            assert (model_config._model_info.default_pooling_type ==
+                    model_info.default_pooling_type)
```
Comment (noooop) [inline tests/models/language/pooling/mteb_utils.py]: Please do not modify this place.

File: tests/models/language/pooling/test_st_projector.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+from ...utils import EmbedModelInfo
+from .mteb_utils import MTEB_EMBED_TOL, mteb_test_embed_models
+
+# ST models with projector (Dense) layers
+ST_PROJECTOR_MODELS = [
+    EmbedModelInfo(
+        "TencentBAC/Conan-embedding-v1",
+        architecture="BertModel",
+        enable_test=True,
+        default_pooling_type="MEAN",
```
Comment (noooop) [inline tests/models/language/pooling/test_st_projector.py]: default_pooling_type  is just for testing cls pooling and last pooling
Please do not set default_pooling_type

**Review by noooop — COMMENTED** [review]
It looks much better now.

File: tests/models/language/pooling/test_st_projector.py (hunk 2 above)
Comment (FFFfff1FFFfff) [reply tests/models/language/pooling/test_st_projector.py]: Gotcha, I'll handle it. Thank you!

Comment (noooop) [reply tests/models/language/pooling/test_st_projector.py]: loading sentence-transformers config has been tested here

https://github.com/vllm-project/vllm/blob/53415653ff24be03e7c90f5b42ef9cb3f72aad71/tests/test_config.py#L229-L242


The best way to test whether the correct pooling_type is using end-to-end test_embed_models_mteb, so we are not testing pooling_type here.


default_pooling_type is for testing whether  @default_pooling_type() is used correctly, e.g.

https://github.com/vllm-project/vllm/blob/53415653ff24be03e7c90f5b42ef9cb3f72aad71/vllm/model_executor/models/bert.py#L328-L332

File: tests/models/language/pooling/mteb_utils.py (hunk 1 above)
Comment (FFFfff1FFFfff) [reply tests/models/language/pooling/mteb_utils.py]: Hi @noooop ,
Since the assertion requires default_pooling_type to match, would this approach work for ST projector tests:

1. Set default_pooling_type="CLS" (for the assertion)
2. Use override_pooler_config=PoolerConfig(pooling_type="MEAN")

Is this reasonable or do you prefer another solution?

Comment (noooop) [reply tests/models/language/pooling/mteb_utils.py]: using CLSPoolingEmbedModelInfo  

default_pooling_type is for testing whether @default_pooling_type() is used correctly

Is the name CLSPoolingEmbedModelInfo not good? Can cause misunderstanding

File: vllm/model_executor/layers/pooler.py [hunk 3]
```diff
@@ -432,12 +436,27 @@ def forward(self, pooled_data: Union[list[torch.Tensor], torch.Tensor],
 
 class EmbeddingPoolerHead(PoolerHead):
 
-    def __init__(self) -> None:
+    def __init__(self, projector: Optional[nn.Module] = None) -> None:
         super().__init__(activation=PoolerNormalize())
+        self.projector = projector
```
Comment (noooop) [inline vllm/model_executor/layers/pooler.py]: how about mv projector=_load_st_projector(vllm_config.model_config) here

This allows us to avoid modifying adapters.py and bert.py, as well as SimplePooler.from_config, and all embedding models can support the Dense module.

Comment (noooop) [reply vllm/model_executor/layers/pooler.py]: try using vllm_config = get_current_vllm_config(), to avoid needing to pass vllm_config in

File: vllm/model_executor/models/adapters.py [hunk 4]
```diff
@@ -24,6 +30,96 @@
...
+        default_weight_loader)
+
+    for filename in ["model.safetensors", "pytorch_model.bin"]:
+        file_path = f"{folder}/{filename}" if folder else filename
+
+        try:
+            file_bytes = get_hf_file_bytes(file_path, model_config.model,
+                                           model_config.revision)
+            if not file_bytes:
+                continue
```
Comment (noooop) [inline vllm/model_executor/models/adapters.py:91]: I don't think we need this

**Comment by noooop** [comment]
(Please make the changes in this PR instead of submitting a new one

File: vllm/model_executor/models/adapters.py [hunk 5]
```diff
@@ -24,6 +30,96 @@
...
+                           layer_config.get("out_features", 768),
+                           bias=layer_config.get("bias", True),
+                           dtype=torch.float32)
+
+        if _load_dense_weights(linear, folder, model_config):
+            layers = [linear]
+            if act_name := layer_config.get("activation_function"):
+                layers.append(get_act_fn(act_name))
+            return nn.Sequential(*layers).to(dtype=torch.float32)
+
```
Comment (noooop) [inline vllm/model_executor/models/adapters.py:71]: According to my experience, the difference in precision between fp16 and fp32 for mlp is not very significant. 

May I ask if setting fp32 makes a difference of 1e-4 in the mteb_test_embed_models test?

File: tests/models/language/pooling/test_st_projector.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+from typing import Any
+
+import pytest
+
+from ...utils import EmbedModelInfo
+from .mteb_utils import MTEB_EMBED_TOL, mteb_test_embed_models
+
+# ST models with projector (Dense) layers
+ST_PROJECTOR_MODELS = [
+    EmbedModelInfo(
```
Comment (noooop) [inline tests/models/language/pooling/test_st_projector.py]: Please use CLSPoolingEmbedModelInfo

File: tests/models/language/pooling/test_st_projector.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+@pytest.mark.parametrize("model_info", ST_PROJECTOR_MODELS)
+def test_embed_models_mteb(hf_runner, vllm_runner,
+                           model_info: EmbedModelInfo) -> None:
+    vllm_extra_kwargs: dict[str, Any] = {}
+
+    mteb_test_embed_models(hf_runner,
+                           vllm_runner,
+                           model_info,
+                           vllm_extra_kwargs,
+                           atol=MTEB_EMBED_TOL)
```
Comment (noooop) [inline tests/models/language/pooling/test_st_projector.py]: vllm_extra_kwargs and atol=MTEB_EMBED_TOL are not needed to be set

File: vllm/model_executor/models/adapters.py [hunk 8]
```diff
@@ -123,11 +219,12 @@ def _init_pooler(self, vllm_config: "VllmConfig", prefix: str = ""):
...
-                {
-                    "encode": Pooler.for_encode(pooler_config),
-                    "embed": Pooler.for_embed(pooler_config),
-                }, )
+            self.pooler = DispatchPooler({
+                "encode":
+                Pooler.for_encode(pooler_config),
+                "embed":
+                Pooler.for_embed(pooler_config),
+            })
```
Comment (noooop) [inline vllm/model_executor/models/adapters.py]: Revert unnecessary changes

File: vllm/model_executor/layers/pooler.py (hunk 3 above)
Comment (FFFfff1FFFfff) [reply vllm/model_executor/layers/pooler.py]: Thank you for the suggestion! @noooop  I've updated the PR accordingly:

- Reverted bert.py
- Moved _load_st_projector from adapter.py to pooler.py
- Updated test_st_projector.py

BTW: I have to add default_pooling_type="CLS" to satisfy the assertion in mteb_utils.py that requires this field to match the model's decorator value. But if you have better idea, please tell me. Thanks!!

**Review by noooop — COMMENTED** [review]
Some minor changes

File: vllm/model_executor/models/adapters.py [hunk 9]
```diff
@@ -24,6 +30,96 @@
...
+
+def _load_dense_weights(linear: nn.Linear, folder: str,
+                        model_config: "ModelConfig") -> bool:
+    """Load weights using vLLM's weight_loader pattern."""
+    from vllm.model_executor.model_loader.weight_utils import (
+        default_weight_loader)
+
+    for filename in ["model.safetensors", "pytorch_model.bin"]:
+        file_path = f"{folder}/{filename}" if folder else filename
+
```
Comment (noooop) [inline vllm/model_executor/models/adapters.py:86]: I think _load_dense_weights needs to be modified, but I don't know how.

File: tests/models/language/pooling/test_st_projector.py (hunk 6 above)
Comment (FFFfff1FFFfff) [reply tests/models/language/pooling/test_st_projector.py]: Oh! I got it now.

**Comment by noooop** [comment]
I can't see any more places that need to modify.

Thanks for the contribution.

cc @DarkLight1337

**Comment by FFFfff1FFFfff** [comment]
> I can't see any more places that need to modify.
> 
> Thanks for the contribution.
> 
> cc @DarkLight1337

Thank you for the review and suggestion! Everything's been updated.
And  @DarkLight1337 please feel free to share any comments or propose changes. 
Thanks a lot!

File: vllm/model_executor/models/adapters.py [hunk 10]
```diff
@@ -24,6 +30,96 @@
...
+                           dtype=torch.float32)
+
+        if _load_dense_weights(linear, folder, model_config):
+            layers = [linear]
+            if act_name := layer_config.get("activation_function"):
+                layers.append(get_act_fn(act_name))
+            return nn.Sequential(*layers).to(dtype=torch.float32)
+
+    except Exception as e:
+        logger.debug("ST projector loading failed: %s", e)
```
Comment (DarkLight1337) [inline vllm/model_executor/models/adapters.py]:
```suggestion
        logger.exception("ST projector loading failed")
```

File: vllm/model_executor/models/adapters.py [hunk 11]
```diff
@@ -24,6 +30,96 @@
...
+
+                    bias_key = weight_key.replace("weight", "bias")
+                    if linear.bias is not None and bias_key in state_dict:
+                        bias_loader = getattr(linear.bias, "weight_loader",
+                                              default_weight_loader)
+                        bias_loader(linear.bias,
+                                    state_dict[bias_key].to(torch.float32))
+                    return True
+        except Exception as e:
+            logger.debug("Failed to load %s: %s", filename, e)
```
Comment (DarkLight1337) [inline vllm/model_executor/models/adapters.py]:
```suggestion
            logger.exception("Failed to load %s", filename)
```

File: vllm/model_executor/models/adapters.py (hunk 10 above)
Comment (DarkLight1337) [reply vllm/model_executor/models/adapters.py]: Better to show the error to the user more clearly

File: vllm/transformers_utils/config.py [hunk 12]
```diff
@@ -927,3 +927,33 @@ def get_model_path(model: Union[str, Path], revision: Optional[str] = None):
...
+        except (OSError, ValueError) as e:
+            logger.debug("Failed to download %s from HF: %s", file_name, e)
+            return None
+
+    if file_path is not None and file_path.is_file():
+        try:
+            with open(file_path, 'rb') as file:
+                return file.read()
+        except OSError as e:
+            logger.debug("Failed to read file %s: %s", file_path, e)
```
Comment (DarkLight1337) [inline vllm/transformers_utils/config.py]: I think no need to soft fail here. Better to just raise the error

Comment (FFFfff1FFFfff) [reply vllm/transformers_utils/config.py]: Thanks for the review! I've addressed all the logging improvements you suggested. Please take a look.

File: vllm/transformers_utils/config.py [hunk 13]
```diff
@@ -927,3 +927,29 @@ def get_model_path(model: Union[str, Path], revision: Optional[str] = None):
...
+
+    if file_path is None:
+        try:
+            hf_hub_file = hf_hub_download(model,
+                                          file_name,
+                                          revision=revision,
+                                          token=_get_hf_token())
+            file_path = Path(hf_hub_file)
+        except (OSError, ValueError) as e:
+            logger.debug("Failed to download %s from HF: %s", file_name, e)
```
Comment (DarkLight1337) [inline vllm/transformers_utils/config.py]: Also should raise the error here

File: vllm/transformers_utils/config.py [hunk 14]
```diff
@@ -927,3 +927,29 @@ def get_model_path(model: Union[str, Path], revision: Optional[str] = None):
...
+
+    if file_path is None:
+        try:
+            hf_hub_file = hf_hub_download(model,
+                                          file_name,
+                                          revision=revision,
+                                          token=_get_hf_token())
+            file_path = Path(hf_hub_file)
+        except Exception:
+            logger.exception("Failed to download %s from HF", file_name)
```
Comment (DarkLight1337) [inline vllm/transformers_utils/config.py]: I mean that we should not catch the exception at all here

Comment (FFFfff1FFFfff) [reply vllm/transformers_utils/config.py]: Got it, I'll get it fixed.

**Review by DarkLight1337 — APPROVED** [review]
LGTM now, thanks!
